### DISCUSSION TITLE
BME280 (180/680): report pressure altitude

### DISCUSSION THREAD

**Discussion by Alexxor36** [arendst/Tasmota#11113] [Q&A] [answered]
Feature request:
Can we add "Pressure Altitude" as a returned value from the BM280 (and related) barometers?

Affected file would be 
Tasmota/tasmota/xsns_09_bmp.ino

Pressure altitude is the altitude one would be at for a given pressure. Of course, as weather changes, the pressure altitude fluctuates for a given station/site. This is a critical factor for aviation weather stations, as hot weather can make a plane 'feel' (and perform) like it was operating at a higher equivalent altitude.

The calculation is already implemented in other BME libraries, eg see Adafruit, lines 514-543:
https://github.com/adafruit/Adafruit_BME280_Library/blob/master/Adafruit_BME280.cpp

And further background:
https://forums.adafruit.com/viewtopic.php?f=22&t=58064

https://en.wikipedia.org/wiki/Pressure_altitude

Note: the adafruit library allows one to correct the pressure altitude by recalibrating with a known altitude at the sensor location. This is not particularly useful, and would require user configuration/input. This fetaure request is for only reporting Pressure Altitude; helpful in airfield weather stations (and people interested in weather).

Alex

**Comment by Jason2866** [comment, marked as answer]
It is there. Provide your altitude with command `Altitude` -30000..30000 = altitude in meters

**Comment by barbudor** [comment]
As far as I understand `float Adafruit_BME280::readAltitude(float seaLevel)` computes the local altitude given
- the local pressure measured by the sensor
- a reference sea level pressure for the same time-of-day

Which means that for this function to report the altitude it needs to be given the reference sealevel pressure at the same time (as the sealevel pressure can change also from weather conditions).

As of today, the Tasmota implemention is working the otherway, calculating and returning the Sea Level pressure given the configured [Altitude](https://tasmota.github.io/docs/Commands/#altitude).

Regarding `float Adafruit_BME280::seaLevelForAltitude(float altitude, float atmospheric)`, this function is already implement in Tasmota as `ConvertPressureForSeaLevel()`. It is used to compute the returned Sea Level Pressure.


Or may be I didn't understood at all your use case.

**Reply by Alexxor36** [reply]
If I type "altitude 1610", I get seapressure 995hPA. (which makes sense - my station is at ~1610m).
What I'm asking is the opposite - something along the lines of 
I input "Seapressure 1016.4" and the return is an altitude number (that will fluctuate as pressure fluctuates). That is called pressure altitude, and would fluctuate around 1610m, but as weather changes (eg hot, low pressure days), it may be as high as ~1800m; which is what an aircraft 'feels' and what is used to compute performance characteristics during flight (and in particular, takeoff performance!).

Alex

**Reply by barbudor** [reply]
Let's see if we can achieve this without modifying Tasmota with rules or script

Right, Tasmota reads the LocalPressure from the BME.
If you have configured the LocalAltitude, Tasmota returns both the LocalPressure and the calculated SeaLevelPressure

Now you would like to enter another AlternateSeaLevelPressure and calculate the NewAltitude that would have the LocalPressure if SealLevelPressure was AlternateSeaLevelPressure ?
Would the AlternatSeaLevelPressure be constant ? or change rarely ?

If you can confirm the formula, I can help you implementing this with RULES + EXPRESSION

**Reply by Alexxor36** [reply]
yes, that makes sense, and AltSeaLevel is a fixed constant = 101325 Pa
(it's a standard, so it is defined as that and does not change)
https://en.wikipedia.org/wiki/Standard_sea-level_conditions

**Reply by barbudor** [reply]
If I take the formula from the Adafruit code you want:

NewAltitude = 44330 * (1 - (MeasuredPressure/1013.25)^0.1903)

What kind of accuracy do you need ?
What is the expected range of local pressure ?

**Reply by Alexxor36** [reply]
accuracy is within whatever the sensor reports - I'm currently actually using the adafruit script (with sealevel set at 101325) to report altitude; but I want the OTA updates, and, at least from my perspective, perceived higher reliability of the tasmota over the arduino script (after a few days, 8266/BME/MQTT arduino script will hang, and need random reboots, which is why I'm trying to do this with tasmota in the first place).

TLDR: whatever the sensor and that equation reports is fine (would be nice if I could convert it to feet, rather than meters, but that should be relatively trivial [ meters * 3.2808]. It's more about the relative change day to day than the absolute accuracy that matters.

Alex

**Reply by barbudor** [reply]
My question was related to the accuracy of the calculation, not the sensor.
I've implemented the function above in a rule to discover major deviation from the same result in C or LO.Calc
The reason is that Tasmota is not using the standard `pow()` function from the C library but a cpu-optimized one which accuracy is poor.

You can test the following rule :
```
Rule1 
   on tele-BMExxx#Pressure do var1 = (44330.0*(1.0-(%value%^0.1903)*0.267924)) endon
   on var1#state do publish tele/%topic%/SENSOR2 {"PressureAltitude":%value%"} endon
```
Replace `BMExxx` by your sensor name and adapt the MQTT message as you whish
Rule need to be enabled by `Rule1 1`
If you have a local display, you may use also the `DisplayText` command instead of MQTT publication.

You will see that values are far from expectations

As the expected BME's pressure is in a limited range I have tried to tweak a better formula using linear approximation for pressures within the range 650 hPA to 1050 hPa:
```
Rule1 
   on tele-BMExxx#Pressure do var1 = (9854.44-9.72613*%value%) endon
   on var1#state do publish tele/%topic%/SENSOR2 {"PressureAltitude":%value%"} endon
```
Better but still can have up to 10% error on the returned altitude when pressure is around 840 hPA.

I hope you get the idea.
If this doesn't suits you, you will have to do the calculation on the back-end upon reception of the original SENSOR message.
For example implementing this in Node-Red is feasible; Probably also in home automation softwares.

**Reply by barbudor** [reply]
@Alexxor36 any feedback on the proposed rules ?
